### ISSUE TITLE
Windows module validation on puppet 4 regressions

### ISSUE THREAD

**Issue by james-powis** [rodjek/rspec-puppet#437]
windows absolute file paths have been broken since 2.3.0, While there are numerous other bug reports around this issue, I figured I would share the results of my Root Cause Analysis and some of the extreme levels of hackery I have attempted to work around the issue only to have future releases cause even more hackery needed. In short the absolute filepath issue was introduced in this commit:

https://github.com/rodjek/rspec-puppet/commit/9c295b5de726128bc08acfb1efaeaf5987174563 

I don't know WHY or how to fix it at this point, so I will leave that to those of you who have more knowledge in this code base than I do.

and Now for your entertainment:

**The Investigation (backstory and hackery)**

On versions after `2.2.0` rspec puppet has required stubbing the file path to avoid getting file path is not fully qualified errors:

example of error:
```log
     Puppet::ResourceError:
       Parameter path failed on File[C:\Windows\DevOps\mssql_std_2014.zip]: File paths must be fully qualified, not 'C:\Windows\DevOps\mssql_std_2014.zip' at /Users/jrp/git/puppet/modules/mssql/spec/fixtures/modules/mssql/manifests/init.pp:44
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/util/errors.rb:106:in `fail'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/type/file.rb:56:in `block (3 levels) in <top (required)>'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/parameter.rb:456:in `validate'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/parameter.rb:491:in `value='
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/type.rb:675:in `[]='
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/type.rb:2454:in `set_name'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/type.rb:2385:in `initialize'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/type/file.rb:491:in `initialize'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/resource.rb:489:in `new'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/resource.rb:489:in `to_ral'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/bundler/gems/rspec-puppet-85bcc190777b/lib/rspec-puppet/matchers/create_generic.rb:262:in `block in relationship_refs'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/type.rb:2071:in `block in eachautorequire'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/type.rb:2070:in `each'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/type.rb:2070:in `eachautorequire'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/bundler/gems/rspec-puppet-85bcc190777b/lib/rspec-puppet/matchers/create_generic.rb:261:in `relationship_refs'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/bundler/gems/rspec-puppet-85bcc190777b/lib/rspec-puppet/matchers/create_generic.rb:256:in `block in relationship_refs'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/bundler/gems/rspec-puppet-85bcc190777b/lib/rspec-puppet/matchers/create_generic.rb:254:in `each'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/bundler/gems/rspec-puppet-85bcc190777b/lib/rspec-puppet/matchers/create_generic.rb:254:in `relationship_refs'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/bundler/gems/rspec-puppet-85bcc190777b/lib/rspec-puppet/matchers/create_generic.rb:256:in `block in relationship_refs'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/bundler/gems/rspec-puppet-85bcc190777b/lib/rspec-puppet/matchers/create_generic.rb:254:in `each'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/bundler/gems/rspec-puppet-85bcc190777b/lib/rspec-puppet/matchers/create_generic.rb:254:in `relationship_refs'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/bundler/gems/rspec-puppet-85bcc190777b/lib/rspec-puppet/matchers/create_generic.rb:284:in `block (2 levels) in precedes?'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/bundler/gems/rspec-puppet-85bcc190777b/lib/rspec-puppet/matchers/create_generic.rb:282:in `each'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/bundler/gems/rspec-puppet-85bcc190777b/lib/rspec-puppet/matchers/create_generic.rb:282:in `block in precedes?'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/bundler/gems/rspec-puppet-85bcc190777b/lib/rspec-puppet/matchers/create_generic.rb:281:in `each'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/bundler/gems/rspec-puppet-85bcc190777b/lib/rspec-puppet/matchers/create_generic.rb:281:in `precedes?'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/bundler/gems/rspec-puppet-85bcc190777b/lib/rspec-puppet/matchers/create_generic.rb:212:in `block in check_requires'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/bundler/gems/rspec-puppet-85bcc190777b/lib/rspec-puppet/matchers/create_generic.rb:211:in `each'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/bundler/gems/rspec-puppet-85bcc190777b/lib/rspec-puppet/matchers/create_generic.rb:211:in `check_requires'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/bundler/gems/rspec-puppet-85bcc190777b/lib/rspec-puppet/matchers/create_generic.rb:100:in `matches?'
     # ./spec/classes/init_spec.rb:43:in `block (4 levels) in <top (required)>'
     # ------------------
     # --- Caused by: ---
     # Puppet::Error:
     #   File paths must be fully qualified, not 'C:\Windows\DevOps\mssql_std_2014.zip'
     #   /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/util/errors.rb:106:in `fail'
```

Example of stub which bypasses this check:
```ruby
      before :each do
        Puppet[:autosign] = false
        Puppet::Util::Platform.stubs(:windows?).returns true
      end
```

This stub has been tested good on versions 2.2.0 - 2.4.0, but becomes non-effected at version 2.5.0 

There is a catch though... if you reply on puppetlabs/powershell then on versions 2.3.0+ you will see this error:

```log
     Puppet::ResourceError:
       Parameter provider failed on Exec[unzip C:\Windows\DevOps\mssql_std_2014.zip]: Invalid exec provider 'powershell' at /Users/jrp/git/puppet/modules/mssql/spec/fixtures/modules/mssql/manifests/init.pp:51
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/type.rb:1921:in `block (2 levels) in providify'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/parameter.rb:456:in `validate'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/parameter.rb:491:in `value='
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/type.rb:675:in `[]='
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/type.rb:2480:in `block in set_parameters'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/type.rb:2474:in `each'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/type.rb:2474:in `set_parameters'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/type.rb:2389:in `initialize'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/resource.rb:489:in `new'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/resource.rb:489:in `to_ral'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/rspec-puppet-2.4.0/lib/rspec-puppet/matchers/create_generic.rb:262:in `block in relationship_refs'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/type.rb:2071:in `block in eachautorequire'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/type.rb:2070:in `each'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/type.rb:2070:in `eachautorequire'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/rspec-puppet-2.4.0/lib/rspec-puppet/matchers/create_generic.rb:261:in `relationship_refs'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/rspec-puppet-2.4.0/lib/rspec-puppet/matchers/create_generic.rb:256:in `block in relationship_refs'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/rspec-puppet-2.4.0/lib/rspec-puppet/matchers/create_generic.rb:254:in `each'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/rspec-puppet-2.4.0/lib/rspec-puppet/matchers/create_generic.rb:254:in `relationship_refs'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/rspec-puppet-2.4.0/lib/rspec-puppet/matchers/create_generic.rb:284:in `block (2 levels) in precedes?'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/rspec-puppet-2.4.0/lib/rspec-puppet/matchers/create_generic.rb:282:in `each'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/rspec-puppet-2.4.0/lib/rspec-puppet/matchers/create_generic.rb:282:in `block in precedes?'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/rspec-puppet-2.4.0/lib/rspec-puppet/matchers/create_generic.rb:281:in `each'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/rspec-puppet-2.4.0/lib/rspec-puppet/matchers/create_generic.rb:281:in `precedes?'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/rspec-puppet-2.4.0/lib/rspec-puppet/matchers/create_generic.rb:212:in `block in check_requires'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/rspec-puppet-2.4.0/lib/rspec-puppet/matchers/create_generic.rb:211:in `each'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/rspec-puppet-2.4.0/lib/rspec-puppet/matchers/create_generic.rb:211:in `check_requires'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/rspec-puppet-2.4.0/lib/rspec-puppet/matchers/create_generic.rb:100:in `matches?'
     # ./spec/classes/init_spec.rb:41:in `block (4 levels) in <top (required)>'
     # ------------------
     # --- Caused by: ---
     # ArgumentError:
     #   Invalid exec provider 'powershell'
     #   /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/type.rb:1921:in `block (2 levels) in providify'
```

This can be easily circumvented with the following stub:
```ruby
      before :each do
        Puppet[:autosign] = false
        Puppet::Util::Platform.stubs(:windows?).returns true
        Puppet::Type.type(:exec).provide :powershell, :parent => Puppet::Provider::Exec do
          def run(command, check = false)
          end
        end
      end
```

Unfortunately like all good things, 2.5.0 broke this stub method resulting in this error:
```log
     Puppet::Error:
       Could not autoload puppet/type/user: Could not autoload puppet/provider/user/windows_adsi: no implicit conversion of Symbol into String.
       Could not open library 'libkernel32.dylib': dlopen(libkernel32.dylib, 5): image not found
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/ffi-1.9.14/lib/ffi/library.rb:147:in `block in ffi_lib'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/ffi-1.9.14/lib/ffi/library.rb:100:in `map'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/ffi-1.9.14/lib/ffi/library.rb:100:in `ffi_lib'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/util/windows/api_types.rb:258:in `<module:WIN32>'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/util/windows/api_types.rb:179:in `<module:APITypes>'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/util/windows/api_types.rb:4:in `<top (required)>'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/util/windows.rb:16:in `require'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/util/windows.rb:16:in `<module:Windows>'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/util/windows.rb:1:in `<top (required)>'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/provider/user/windows_adsi.rb:1:in `require'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/provider/user/windows_adsi.rb:1:in `<top (required)>'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/util/autoload.rb:68:in `load'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/util/autoload.rb:68:in `load_file'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/util/autoload.rb:83:in `block in loadall'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/util/autoload.rb:81:in `each'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/util/autoload.rb:81:in `loadall'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/util/autoload.rb:208:in `loadall'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/metatype/manager.rb:126:in `newtype'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/type/user.rb:9:in `<module:Puppet>'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/type/user.rb:8:in `<top (required)>'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/util/autoload.rb:68:in `load'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/util/autoload.rb:68:in `load_file'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/util/autoload.rb:194:in `load'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/metatype/manager.rb:171:in `type'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/resource.rb:390:in `resource_type'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/resource.rb:378:in `resource_type'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/resource.rb:319:in `initialize'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/resource/catalog.rb:366:in `new'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/resource/catalog.rb:366:in `resource'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/rspec-puppet-2.5.0/lib/rspec-puppet/matchers/create_generic.rb:239:in `resource_from_ref'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/rspec-puppet-2.5.0/lib/rspec-puppet/matchers/create_generic.rb:243:in `canonicalize_resource'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/rspec-puppet-2.5.0/lib/rspec-puppet/matchers/create_generic.rb:251:in `relationship_refs'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/rspec-puppet-2.5.0/lib/rspec-puppet/matchers/create_generic.rb:264:in `block (2 levels) in relationship_refs'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/rspec-puppet-2.5.0/lib/rspec-puppet/matchers/create_generic.rb:262:in `each'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/rspec-puppet-2.5.0/lib/rspec-puppet/matchers/create_generic.rb:262:in `block in relationship_refs'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/type.rb:2071:in `block in eachautorequire'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/type.rb:2070:in `each'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/puppet-4.8.1-universal-darwin/lib/puppet/type.rb:2070:in `eachautorequire'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/rspec-puppet-2.5.0/lib/rspec-puppet/matchers/create_generic.rb:261:in `relationship_refs'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/rspec-puppet-2.5.0/lib/rspec-puppet/matchers/create_generic.rb:256:in `block in relationship_refs'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/rspec-puppet-2.5.0/lib/rspec-puppet/matchers/create_generic.rb:254:in `each'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/rspec-puppet-2.5.0/lib/rspec-puppet/matchers/create_generic.rb:254:in `relationship_refs'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/rspec-puppet-2.5.0/lib/rspec-puppet/matchers/create_generic.rb:284:in `block (2 levels) in precedes?'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/rspec-puppet-2.5.0/lib/rspec-puppet/matchers/create_generic.rb:282:in `each'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/rspec-puppet-2.5.0/lib/rspec-puppet/matchers/create_generic.rb:282:in `block in precedes?'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/rspec-puppet-2.5.0/lib/rspec-puppet/matchers/create_generic.rb:281:in `each'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/rspec-puppet-2.5.0/lib/rspec-puppet/matchers/create_generic.rb:281:in `precedes?'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/rspec-puppet-2.5.0/lib/rspec-puppet/matchers/create_generic.rb:212:in `block in check_requires'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/rspec-puppet-2.5.0/lib/rspec-puppet/matchers/create_generic.rb:211:in `each'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/rspec-puppet-2.5.0/lib/rspec-puppet/matchers/create_generic.rb:211:in `check_requires'
     # /Users/jrp/.rvm/gems/ruby-2.3.0/gems/rspec-puppet-2.5.0/lib/rspec-puppet/matchers/create_generic.rb:100:in `matches?'
     # ./spec/classes/init_spec.rb:41:in `block (4 levels) in <top (required)>'
     # ------------------
     # --- Caused by: ---
     # LoadError:
     #   no implicit conversion of Symbol into String.
     #   Could not open library 'libkernel32.dylib': dlopen(libkernel32.dylib, 5): image not found
     #   /Users/jrp/.rvm/gems/ruby-2.3.0/gems/ffi-1.9.14/lib/ffi/library.rb:147:in `block in ffi_lib'
```

Here are the commit ID's which Introduced each of the issues identified at version numbers:
2.5.0: https://github.com/rodjek/rspec-puppet/commit/d351413ce5c8757fcdd4f69db0e0d89a9cf88c82

2.x with the absolute file path stub: Never worked without the powershell stub too when using provider => 'powershell' in an exec resource, however as mentioned on 2.2.0 everything worked without any stubs.

Working without any stubs broken 2.3.0 as a result of: https://github.com/rodjek/rspec-puppet/commit/9c295b5de726128bc08acfb1efaeaf5987174563 

I think the best approach is to look at what can be done to restore the broken functionality introduced in 2.3 and ignore all the stub foo as I think the stubs themselves are causing some of the issues.

**Comment by DavidS** [comment]
Thanks for your excellent analysis, I'll have a fresh look at this.

**Comment by rodjek** [comment]
This should now be resolved by #498
